algorithm/cgal/cgal.cc

```diff
@@ -4280,65 +4478,97 @@ int Disjoint(Geometry* geometry, emscripten::val getIsMasked) {
-              geometry->plane(start) != geometry->plane(nth) ||
-              geometry->noGpsOverlap(start, nth)) {
+          if (is_masked[nth]) {
             continue;
           }
-          geometry->gps(start).difference(geometry->gps(nth));
-          geometry->updateGpsBounds(start);
+          switch (geometry->getType(nth)) {
+            case GEOMETRY_POLYGONS_WITH_HOLES: {
+              if (geometry->plane(start) != geometry->plane(nth) ||
+                  geometry->noOverlap2(start, nth)) {
+                continue;
+              }
+              geometry->gps(start).difference(geometry->gps(nth));
+              geometry->updateBounds2(start);
+              break;
+            }
+            case GEOMETRY_MESH: {
+              // TODO: Handle disjunction of surface by volume.
+              break;
+            }
+          }
         }
-        // TODO: Handle disjunction of surface by volume.
         break;
       }
       case GEOMETRY_SEGMENTS: {
-        // TODO: Support disjunction by PolygonsWithHoles.
-        std::vector<Segment> in;
-        geometry->input_segments(start).swap(in);
-        std::vector<Segment> out;
         for (int nth = start + 1; nth < size; nth++) {
-          if (is_masked[nth] || !geometry->is_mesh(nth) ||
-              geometry->is_empty_mesh(nth)) {
+          if (is_masked[nth]) {
             continue;
           }
-          if (queries[nth] == nullptr && geometry->is_mesh(nth)) {
-            queries[nth].reset(
-                new SurfaceMeshQuery(&geometry->mesh(nth), &identity));
-          }
-          std::unique_ptr<SurfaceMeshQuery>& query = queries[nth];
-          auto emit = [&](const Segment& segment) { out.push_back(segment); };
-          for (const Segment& segment : in) {
-            query->intersectSegment(false, segment, emit);
+          switch (geometry->getType(nth)) {
+            case GEOMETRY_MESH: {
+              if (query[nth] == nullptr) {
+                query[nth].reset(new SurfaceMeshQuery(geometry->mesh(nth)));
+              }
+              Segments out;
+              // TODO: See if we can leverage std::back_inserter instead of an
+              // lexical closure.
+              auto emit = [&](const Segment& segment) {
+                out.push_back(segment);
+              };
+              for (const Segment& segment : geometry->segments(start)) {
+                query[nth]->intersectSegment(false, segment, emit);
+              }
+              geometry->segments(start).swap(out);
+              break;
+            }
+            case GEOMETRY_SEGMENTS: {
+              // TODO: Support disjunction by polygons-with-holes.
+              break;
+            }
           }
-          in.swap(out);
-          out.clear();
         }
-        geometry->segments(start).swap(in);
         break;
       }
       case GEOMETRY_POINTS: {
-        // TBD
+        // TODO: Support disjunction by volumes, segments, polygons.
```

Comment:
This is the ability of a say 3D object to cut a point cloud for example?

Reply:
Yeah. Pretty much everything should be able to cut points.